browser/brave_rewards/creator_detection_script_injector.cc

```diff
@@ -0,0 +1,182 @@
+    return std::nullopt;
+  }
+
+  // Only run scripts if the user has enabled Brave Rewards.
+  if (!profile || !profile->GetPrefs()->GetBoolean(prefs::kEnabled)) {
+    return std::nullopt;
+  }
+
+  // Only run scripts for known "media platform" sites.
+  GURL url = rfh->GetLastCommittedURL();
+  if (!IsMediaPlatformURL(url)) {
+    return std::nullopt;
+  }
+
+  // Only run scripts when there is an exact hostname match.
+  auto iter = kScriptMap.find(url.host_piece());
+  if (iter == kScriptMap.end()) {
+    return std::nullopt;
+  }
+
+  return LoadScriptResource(iter->second);
+}
+
+}  // namespace
+
+CreatorDetectionScriptInjector::CreatorDetectionScriptInjector() = default;
+CreatorDetectionScriptInjector::~CreatorDetectionScriptInjector() = default;
+
+void CreatorDetectionScriptInjector::MaybeInjectScript(
+    content::RenderFrameHost* rfh) {
+  injector_.reset();
+  injector_host_token_ = content::GlobalRenderFrameHostToken();
+
+  if (!rfh) {
+    return;
+  }
+
+  auto script_source = GetDetectionScript(rfh);
+  if (!script_source) {
+    return;
+  }
+
+  injector_host_token_ = rfh->GetGlobalFrameToken();
+  rfh->GetRemoteAssociatedInterfaces()->GetInterface(&injector_);
+
+  bool verbose_logging =
+      base::FeatureList::IsEnabled(features::kVerboseLoggingFeature);
+
+  // Add a `braveRewards` global variable to the isolated world.
+  std::string initializer =
+      base::StrCat({"self.braveRewards = { verboseLogging: ",
+                    verbose_logging ? "true" : "false", " }"});
+  ExecuteScript(initializer, base::DoNothing());
+
+  // Execute the detection script. It must set `braveRewards.detectCreator` to a
+  // function. That function will be called by `DetectCreator`.
+  ExecuteScript(script_source.value(), base::DoNothing());
+}
+
+void CreatorDetectionScriptInjector::DetectCreator(
+    content::RenderFrameHost* rfh,
+    DetectCreatorCallback callback) {
+  if (!rfh || rfh->GetGlobalFrameToken() != injector_host_token_ ||
+      !injector_.is_bound()) {
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+
+  // Call the detection function set up by the detection script.
+  ExecuteScript(
+      "braveRewards.detectCreator()",
+      base::BindOnce(&CreatorDetectionScriptInjector::OnCreatorDetected,
+                     weak_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void CreatorDetectionScriptInjector::ExecuteScript(
+    std::string_view script,
+    ExecuteScriptCallback callback) {
+  CHECK(injector_.is_bound());
+  injector_->RequestAsyncExecuteScript(
+      ISOLATED_WORLD_ID_BRAVE_INTERNAL, base::UTF8ToUTF16(script),
+      blink::mojom::UserActivationOption::kDoNotActivate,
+      blink::mojom::PromiseResultOption::kAwait, std::move(callback));
+}
+
+void CreatorDetectionScriptInjector::OnCreatorDetected(
+    DetectCreatorCallback callback,
+    base::Value value) {
+  Result result;
+  if (auto* dict = value.GetIfDict()) {
+    if (auto* id = dict->FindString("id")) {
```

Comment:
is it ok for some/all of these to be null? Does something else validate it later?

Reply:
It is OK. `RewardsTabHelper` checks the "id" member to determine whether creator info was detected. Each of the other members is best-effort.